kategory-effects/src/main/kotlin/kategory/effects/data/DeferredKW.kt

```diff
@@ -0,0 +1,110 @@
+package kategory
+
+import kategory.effects.data.internal.Platform
+import kategory.effects.instances.DeferredKWInstances
+import kotlinx.coroutines.experimental.*
+import kotlin.coroutines.experimental.CoroutineContext
+
+typealias DeferredResult<A> = Either<Throwable, A>
+
+@higherkind data class DeferredKW<out A>(val coroutineContext: CoroutineContext, val thunk: (CoroutineContext) -> Deferred<A>) : DeferredKWKind<A> {
+
+    fun runDeferred(): Deferred<A> =
+            thunk(coroutineContext)
+
+    fun <B> map(f: (A) -> B): DeferredKW<B> =
+            flatMap { a: A -> pure(coroutineContext, f(a)) }
+
+    fun <B> flatMap(f: (A) -> DeferredKW<B>): DeferredKW<B> =
+            DeferredKW(coroutineContext) {
+                this@DeferredKW.attempt().fold(
+                        { raiseError<B>(coroutineContext, it).runDeferred() },
+                        { f(it).runDeferred() })
+            }
+
+    companion object {
+        inline operator fun <A> invoke(coroutineContext: CoroutineContext, noinline f: suspend CoroutineScope.() -> A): DeferredKW<A> =
+                DeferredKW(coroutineContext) { context: CoroutineContext ->
+                    Platform.onceOnly<Unit> { Unit }.let {
+                        async(context, CoroutineStart.DEFAULT) {
+                            f()
+                        }
+                    }
+                }
+
+        inline fun <A> async(coroutineContext: CoroutineContext, noinline fa: Proc<A>): DeferredKW<A> =
+                DeferredKW(coroutineContext) { context: CoroutineContext ->
+                    Platform.onceOnly<Unit> { Unit }.let {
+                        async(context, CoroutineStart.DEFAULT) {
+                            IO.async(fa).unsafeRunSync()
+                        }
+                    }
+                }
+
+        inline fun <A> pure(coroutineContext: CoroutineContext, a: A): DeferredKW<A> =
+                DeferredKW.invoke(coroutineContext, { a })
+
+        inline fun <A> raiseError(coroutineContext: CoroutineContext, t: Throwable): DeferredKW<A> =
+                DeferredKW(coroutineContext) { throw t }
+
+        fun <A, B> tailRecM(coroutineContext: CoroutineContext, a: A, f: (A) -> DeferredKW<Either<A, B>>): DeferredKW<B> {
+            tailrec fun go(a: A, f: (A) -> DeferredKW<Either<A, B>>): DeferredKW<B> {
+                val result: DeferredResult<Either<A, B>> = f(a).attempt()
```

Comment:
We should not block to obtain the result in `tailRecM`. Given the right async context his monad is already stacksafe so an implementation of `tailRecM` here should just use the regular combinators and there should be no need to declare this function as `tailrec`.
Something like this should work for all stack-safe monads:
```kotlin
fun <F, A, B> tailRecM(a: A, f: A -> HK<F, Either<A, B>>): HK<F, B> = flatMap(f(a), {
    when (it) {
      is Left => tailRecM(it.a, f)
      is Right => pure(it.b)
    }
}
```

Reply:
That version it's orders of magnitude slower to the point of looking like a freeze, I'm not sure why :/